File: yeonddori/2024-05/BOJ_4949_균형잡힌-세상.py
text = input()
while text != '.':
    stack = []
    answer = True
    for char in text:
        if char == '(' or char == '[':
            stack.append(char)
        elif char == ')':
            if not stack or stack[-1] == '[':
                answer = False
                break
            stack.pop()
        elif char == ']':
            if not stack or stack[-1] == '(':
                answer = False
                break
            stack.pop()
    if stack:
        answer = False
    print('yes' if answer else 'no')
    text = input()

File: banchan01/2024-06/BOJ_4949_균형잡힌세상.cpp
#include <iostream>
#include <string>
#include <stack>

using namespace std;

int main()
{
    ios::sync_with_stdio(0);
    cin.tie(0);

    while(1)
    {
        bool flag = true;
        stack<int> st;
        // 입력받기
        string s;
        getline(cin,s);

        // 종료조건
        if(s == ".")
            break;

        // 괄호 스택에 넣고, 지우기
        for(char ch : s)
        {
            if(ch=='(' || ch=='[')
            {
                st.push(ch);
            }

            else if(ch==')')
            {
                if(st.empty() || st.top()!='(')
                {
                    flag = false;
                    break;
                }
                st.pop();
            }
            else if(ch==']')
            {
                if(st.empty() || st.top()!='[')
                {
                    flag = false;
                    break;
                }
                st.pop();
            }
            
        }
        if(!st.empty())
            flag = false;
        if(flag)
            cout << "yes" << '\n';
        else
            cout << "no" << '\n';
                
    }
}
